File: gke_customization/gke_catalog/report/batch_dashboard/batch_dashboard.py
# Copyright (c) 2026, Gurukrupa Export and contributors
# For license information, please see license.txt

import frappe
from frappe import _


def execute(filters=None):
    filters = filters or {}
    columns = get_columns()
    data = get_data(filters)
    return columns, data, None, None, None


def get_columns():
    return [
        {"label": _("MWO"), "fieldname": "batch_no", "fieldtype": "Link", "options": "Manufacturing Work Order", "width": 140},
        {"label": _("To Department"), "fieldname": "to_department", "fieldtype": "Link", "options": "Department", "width": 130},
        {"label": _("Manager"), "fieldname": "manager", "fieldtype": "Data", "width": 120},
        {"label": _("To Manager"), "fieldname": "to_manager", "fieldtype": "Data", "width": 120},
        {"label": _("Employee"), "fieldname": "employee", "fieldtype": "Data", "width": 120},
        {"label": _("PMO"), "fieldname": "order_no", "fieldtype": "Link", "options": "Parent Manufacturing Order", "width": 150},
        {"label": _("Sales Order"), "fieldname": "sales_order", "fieldtype": "Link", "options": "Sales Order", "width": 150},
        {"label": _("Due Date"), "fieldname": "due_date", "fieldtype": "Date", "width": 110},
        {"label": _("Touch"), "fieldname": "touch", "fieldtype": "Data", "width": 90},
        {"label": _("Dia Type"), "fieldname": "dia_type", "fieldtype": "Data", "width": 120},
        {"label": _("Setting"), "fieldname": "setting", "fieldtype": "Data", "width": 120},
        {"label": _("Style Bio"), "fieldname": "style_bio", "fieldtype": "Data", "width": 130},
        {"label": _("Category"), "fieldname": "category", "fieldtype": "Data", "width": 120},
        {"label": _("Sub Category"), "fieldname": "sub_category", "fieldtype": "Data", "width": 120},
        {"label": _("Status"), "fieldname": "status", "fieldtype": "Data", "width": 120},
        {"label": _("Company"), "fieldname": "company", "fieldtype": "Link", "options": "Company", "width": 120},
    ]


def get_data(filters):
    conditions, values = get_conditions(filters)

    data = frappe.db.sql(
        f"""
        SELECT
            MWO.name AS batch_no,
            MWO.department AS to_department,
            '' AS manager,
            '' AS to_manager,
            '' AS employee,
            PMO.name AS order_no,
            PMO.sales_order AS sales_order,
            MWO.delivery_date AS due_date,
            PMO.metal_purity AS touch,
            PMO.diamond_grade AS dia_type,
            MWO.setting_type AS setting,
            MO.item_code AS style_bio,
            MWO.item_category AS category,
            MWO.item_sub_category AS sub_category,
            MWO.status AS status,
            MWO.company AS company
        FROM `tabManufacturing Work Order` MWO
        LEFT JOIN `tabParent Manufacturing Order` PMO
            ON MWO.manufacturing_order = PMO.name
        LEFT JOIN `tabManufacturing Operation` MO
            ON MWO.manufacturing_operation = MO.name
        WHERE MWO.manufacturing_order IS NOT NULL {conditions}
        AND EXISTS (
            SELECT 1 FROM `tabManufacturing Work Order` MWO2
            WHERE MWO2.manufacturing_order = MWO.manufacturing_order
            AND NOT (MWO2.status = 'Completed' AND MWO2.docstatus = 1)
        )
        ORDER BY MWO.delivery_date ASC, MWO.name ASC
        """,
        values,
        as_dict=1,
    )

    return data


def get_conditions(filters):
    conditions = ""
    values = {}

    if filters.get("sales_order"):
        conditions += " AND PMO.sales_order = %(sales_order)s"
        values["sales_order"] = filters.get("sales_order")

    if filters.get("to_department"):
        conditions += " AND MWO.department = %(to_department)s"
        values["to_department"] = filters.get("to_department")

    if filters.get("status"):
        conditions += " AND MWO.status = %(status)s"
        values["status"] = filters.get("status")

    if filters.get("company"):
        conditions += " AND MWO.company = %(company)s"
        values["company"] = filters.get("company")

    if filters.get("from_due_date"):
        conditions += " AND MWO.delivery_date >= %(from_due_date)s"
        values["from_due_date"] = filters.get("from_due_date")

    if filters.get("to_due_date"):
        conditions += " AND MWO.delivery_date <= %(to_due_date)s"
        values["to_due_date"] = filters.get("to_due_date")

    return conditions, values


@frappe.whitelist()
def get_batch_detail(batch_no):
    rows = frappe.db.sql(
        """
        SELECT
            SED.parent AS stock_entry,
            SED.item_code AS item_code,
            SED.item_name AS item_name,
            SED.s_warehouse AS s_warehouse,
            SED.t_warehouse AS t_warehouse,
            SED.batch_no AS batch_no,
            SED.qty AS qty,
            SED.uom AS uom,
            SED.basic_rate AS basic_rate,
            SED.amount AS amount
        FROM `tabStock Entry Detail` SED
        INNER JOIN `tabStock Entry` SE ON SE.name = SED.parent
        WHERE SE.manufacturing_work_order = %(batch_no)s
        ORDER BY SED.parent ASC, SED.idx ASC
        """,
        {"batch_no": batch_no},
        as_dict=True,
    )
    return rows


@frappe.whitelist()
def get_batch_history(batch_no):
    rows = frappe.db.sql(
    """
    SELECT
        MO.manufacturing_work_order AS batch_no,
        TDIR.previous_department AS from_department,
        TDIR.current_department AS to_department,
        DATE(TDIR.date_time) AS receive_date,
        TIME(TDIR.date_time) AS receive_time,
        MO.employee AS employee_name,
        COALESCE(MO.operation, MO.name) AS current_operation
    FROM `tabManufacturing Operation` MO
    LEFT JOIN `tabDepartment IR Operation` DIRO
        ON DIRO.manufacturing_operation = MO.name
    LEFT JOIN `tabDepartment IR` TDIR
        ON DIRO.parent = TDIR.name
    WHERE MO.manufacturing_work_order = %s
    AND MO.department_ir_status ='Received'
    ORDER BY TDIR.date_time ASC
    """,
    (batch_no,),
    as_dict=True,
)
    return rows

File: gke_customization/gke_catalog/report/batch_dashboard/batch_dashboard.js
// Copyright (c) 2026, Gurukrupa Export and contributors
// For license information, please see license.txt


frappe.query_reports["Batch dashboard"] = {
    filters: [
        {
            fieldname: "sales_order",
            label: __("Sales Order"),
            fieldtype: "Link",
            options: "Sales Order",
        },
        {
            fieldname: "to_department",
            label: __("To Department"),
            fieldtype: "Link",
            options: "Department",
        },
        {
            fieldname: "status",
            label: __("Status"),
            fieldtype: "Select",
            options: [
                "",
                "Draft",
                "Not Started",
                "In Process",
                "Completed",
                "Stopped",
                "Closed",
                "Cancelled",
            ].join("\n"),
        },
        {
            fieldname: "from_due_date",
            label: __("Due Date From"),
            fieldtype: "Date",
        },
        {
            fieldname: "to_due_date",
            label: __("Due Date To"),
            fieldtype: "Date",
        },
    ],

    formatter: function (value, row, column, data, default_formatter) {
        value = default_formatter(value, row, column, data);

        if (column.fieldname === "batch_no" && data && data.batch_no) {
            value = `
                <a href="#" onclick="return false;" class="batch-detail-icon" data-batch="${data.batch_no}" title="${__("Batch Detail")}">
                    <i class="fa fa-search-plus" style="margin-right:6px;"></i>
                </a>
                <a href="#" onclick="return false;" class="batch-history-icon" data-batch="${data.batch_no}" title="${__("Batch History")}">
                    <i class="fa fa-history" style="margin-right:6px;"></i>
                </a>
                ${value}
            `;
        }

        return value;
    },

    onload: function (report) {
        report.page.main.on("click", ".batch-detail-icon", function () {
            const $btn = $(this);
            if ($btn.hasClass("disabled")) return;
            $btn.addClass("disabled");
            show_batch_detail_dialog($btn.data("batch"), () => $btn.removeClass("disabled"));
        });

        report.page.main.on("click", ".batch-history-icon", function () {
            const $btn = $(this);
            if ($btn.hasClass("disabled")) return;
            $btn.addClass("disabled");
            show_batch_history_dialog($btn.data("batch"), () => $btn.removeClass("disabled"));
        });
    },
};

function show_batch_detail_dialog(batch_no, on_close) {
    function to_float(v) {
        return typeof flt === "function" ? flt(v) : parseFloat(v) || 0;
    }

    frappe.call({
        method: "gke_customization.gke_catalog.report.batch_dashboard.batch_dashboard.get_batch_detail",
        args: {
            batch_no: batch_no,
        },
        callback: function (r) {
            const rows = r.message || [];
            const total_qty = rows.reduce((sum, d) => sum + (to_float(d.qty) || 0), 0);
            const total_amount = rows.reduce((sum, d) => sum + (to_float(d.amount) || 0), 0);

            let html = `
                <table class="table table-bordered">
                    <thead><tr>
                        <th>${__("Stock Entry")}</th><th>${__("Item Code")}</th><th>${__("Item Name")}</th>
                        <th>${__("Source Warehouse")}</th><th>${__("Target Warehouse")}</th>
                        <th>${__("Batch No")}</th>
                        <th>${__("Qty")}</th><th>${__("UOM")}</th>
                        <th>${__("Rate")}</th><th>${__("Amount")}</th>
                    </tr></thead>
                    <tbody>`;

            if (!rows.length) {
                html += `
                    <tr>
                        <td colspan="10" class="text-center text-muted">${__("No stock entry found")}</td>
                    </tr>
                `;
            } else {
                rows.forEach((d) => {
                    html += `<tr>
                        <td>${d.stock_entry || ""}</td><td>${d.item_code || ""}</td><td>${d.item_name || ""}</td>
                        <td>${d.s_warehouse || ""}</td><td>${d.t_warehouse || ""}</td>
                        <td>${d.batch_no || ""}</td>
                        <td class="text-right">${to_float(d.qty).toFixed(3)}</td><td>${d.uom || ""}</td>
                        <td class="text-right">${to_float(d.basic_rate).toFixed(2)}</td>
                        <td class="text-right">${to_float(d.amount).toFixed(2)}</td>
                    </tr>`;
                });
            }

            html += `</tbody>
                <tfoot><tr>
                    <th colspan="6" class="text-right">${__("Total")}</th>
                    <th class="text-right">${total_qty.toFixed(3)}</th>
                    <th></th><th></th>
                    <th class="text-right">${total_amount.toFixed(2)}</th>
                </tr></tfoot>
                </table>`;

            new frappe.ui.Dialog({
                title: __("Batch Detail") + " - " + batch_no,
                size: "large",
                fields: [{ fieldtype: "HTML", options: html }],
                onhide: on_close,
            }).show();
        },
        error: function () {
            on_close && on_close();
        },
    });
}

function show_batch_history_dialog(batch_no, on_close) {
    frappe.call({
        method: "gke_customization.gke_catalog.report.batch_dashboard.batch_dashboard.get_batch_history",
        args: {
            batch_no: batch_no,
        },
        callback: function (r) {
            const rows = r.message || [];
            let html = `
                <table class="table table-bordered">
                    <thead>
                        <tr>
                            <th>${__("Batch No")}</th>
                            <th>${__("From Department")}</th>
                            <th>${__("To Department")}</th>
                            <th>${__("Receive Date")}</th>
                            <th>${__("Receive Time")}</th>
                            <th>${__("Employee Name")}</th>
                            <th>${__("Operation")}</th>
                        </tr>
                    </thead>
                    <tbody>
            `;

            if (!rows.length) {
                html += `
                    <tr>
                        <td colspan="7" class="text-center text-muted">${__("No history found")}</td>
                    </tr>
                `;
            } else {
                rows.forEach((d) => {
                    html += `
                        <tr>
                            <td>${d.batch_no || ""}</td>
                            <td>${d.from_department || ""}</td>
                            <td>${d.to_department || ""}</td>
                            <td>${d.receive_date || ""}</td>
                            <td>${d.receive_time || ""}</td>
                            <td>${d.employee_name || ""}</td>
                            <td>${d.current_operation || ""}</td>
                        </tr>
                    `;
                });
            }

            html += `</tbody></table>`;

            new frappe.ui.Dialog({
                title: __("Batch History") + " - " + batch_no,
                size: "extra-large",
                fields: [{ fieldtype: "HTML", options: html }],
                onhide: on_close,
            }).show();
        },
        error: function () {
            on_close && on_close();
        },
    });
}